src/ir/module-utils.cpp

```diff
@@ -216,37 +196,153 @@ std::vector<HeapType> collectHeapTypes(Module& wasm) {
   return types;
 }
 
-IndexedHeapTypes getIndexedHeapTypes(Module& wasm) {
+IndexedHeapTypes getOptimizedIndexedHeapTypes(Module& wasm) {
   Counts counts = getHeapTypeCounts(wasm);
-  IndexedHeapTypes indexedTypes;
+
+  if (getTypeSystem() != TypeSystem::Isorecursive) {
+    // Sort by frequency and then original insertion order.
+    std::vector<std::pair<HeapType, size_t>> sorted(counts.begin(),
+                                                    counts.end());
+    std::stable_sort(sorted.begin(), sorted.end(), [&](auto a, auto b) {
+      return a.second > b.second;
+    });
+
+    // Collect the results.
+    IndexedHeapTypes indexedTypes;
+    for (Index i = 0; i < sorted.size(); ++i) {
+      indexedTypes.types.push_back(sorted[i].first);
+    }
+    setIndices(indexedTypes);
+    return indexedTypes;
+  }
+
+  // Isorecursive types have more ordering constraints than other types.
+  // Specifically, all types in a recursion group must be adjacent, groups must
+  // be ordered after other groups they reference, and supertypes must be
+  // ordered before their subtypes. We want to order by decreasing reference
+  // counts and have a deterministic ordering while respecting those
+  // constraints. To do that, we first order recursion groups by:
+  //
+  //   1. Dependencies between groups
+  //   2. Average reference count of group members
+  //   3. Original appearance order.
+  //
+  // We then order the types within each group by:
+  //
+  //   1. Supertype dependencies
+  //   2. Reference count
+  //   3. Original group index.
+
+  struct GroupInfo {
+    std::unordered_set<RecGroup> predecessors;
+    size_t count;
+    size_t appearanceIndex;
+  };
+
+  InsertOrderedMap<RecGroup, GroupInfo> groupInfos;
   for (auto& [type, _] : counts) {
-    indexedTypes.types.push_back(type);
+    RecGroup group = type.getRecGroup();
+    // Try to initialize a new info or get the existing info.
+    std::pair<const RecGroup, GroupInfo> entry{group,
+                                               {{}, 0, groupInfos.size()}};
+    auto& info = groupInfos.insert(entry).first->second;
+    // Update the reference count.
+    info.count += counts.at(type);
+    // Collect predecessor groups.
+    for (auto child : type.getReferencedHeapTypes()) {
+      RecGroup otherGroup = child.getRecGroup();
+      if (otherGroup != group) {
+        info.predecessors.insert(otherGroup);
+      }
+    }
   }
 
-  coalesceRecGroups(indexedTypes);
-  setIndices(indexedTypes);
-  return indexedTypes;
-}
+  // Fix up the cumulative counts to be an average instead.
+  for (auto& [group, info] : groupInfos) {
+    info.count /= group.size();
```

Comment:
Why average here?

Reply:
Reasonable alternatives might be to take the max or min count among group members instead, but average is a better heuristic.

If we took the max count, then we could end up with huge groups with many infrequently used types and a few commonly used types hogging the valuable beginning of the index space.

If we took the min count, then we could end up with groups with many frequently used types and a few infrequently used types appearing very late in the index space, which would be bad for code size.

Using the average means that the cumulative count for the group has to justify the amount of index space the group takes up.

Reply:
I'll add a comment in the code.

Reply:
I see, makes sense.

The averaging using integers loses information, though. In fact I wonder if it might be common for the average to end up mostly 0 or 1. Maybe using doubles would make sense? The overhead isn't large, and they are deterministic.